diff --git a/dubbo-common/src/main/java/org/apache/dubbo/common/URLStrParser.java b/dubbo-common/src/main/java/org/apache/dubbo/common/URLStrParser.java
index 3b1761e2ff..c8ee6337e4 100644
--- a/dubbo-common/src/main/java/org/apache/dubbo/common/URLStrParser.java
+++ b/dubbo-common/src/main/java/org/apache/dubbo/common/URLStrParser.java
@@ -20,0 +21 @@ import org.apache.dubbo.common.url.component.URLItemCache;
+import org.apache.dubbo.common.utils.StringUtils;
@@ -26,0 +28,2 @@ import static org.apache.dubbo.common.constants.CommonConstants.DEFAULT_KEY_PREF
+import static org.apache.dubbo.common.constants.CommonConstants.PASSWORD_KEY;
+import static org.apache.dubbo.common.constants.CommonConstants.USERNAME_KEY;
@@ -45,2 +48,2 @@ public final class URLStrParser {
-     * @param decodedURLStr : after {@link URL#decode} string
-     *                      decodedURLStr format: protocol://username:password@host:port/path?k1=v1&k2=v2
+     * @param decodedURLStr : after {@link URL#decode} string decodedURLStr format:
+     *                      protocol://username:password@host:port/path?k1=v1&k2=v2
@@ -105,0 +109 @@ public final class URLStrParser {
+
@@ -149,0 +154,9 @@ public final class URLStrParser {
+        if (username == null && parameters != null && StringUtils.isNotEmpty(parameters.get(USERNAME_KEY))) {
+            username = parameters.get(USERNAME_KEY);
+            parameters.remove(USERNAME_KEY);
+        }
+
+        if (password == null && parameters != null && StringUtils.isNotEmpty(parameters.get(PASSWORD_KEY))) {
+            password = parameters.get(PASSWORD_KEY);
+            parameters.remove(PASSWORD_KEY);
+        }
@@ -206,2 +219,2 @@ public final class URLStrParser {
-     * @param encodedURLStr : after {@link URL#encode(String)} string
-     *                      encodedURLStr after decode format: protocol://username:password@host:port/path?k1=v1&k2=v2
+     * @param encodedURLStr : after {@link URL#encode(String)} string encodedURLStr after decode format:
+     *                      protocol://username:password@host:port/path?k1=v1&k2=v2
diff --git a/dubbo-common/src/test/java/org/apache/dubbo/common/URLStrParserTest.java b/dubbo-common/src/test/java/org/apache/dubbo/common/URLStrParserTest.java
index 4c4c218c35..01364faab9 100644
--- a/dubbo-common/src/test/java/org/apache/dubbo/common/URLStrParserTest.java
+++ b/dubbo-common/src/test/java/org/apache/dubbo/common/URLStrParserTest.java
@@ -50,0 +51 @@ class URLStrParserTest {
+        testCases.add("zookeeper://test10.301.216.302:2181?username=t1#&password=t2#");
