diff --git a/dubbo-common/src/main/java/org/apache/dubbo/common/URL.java b/dubbo-common/src/main/java/org/apache/dubbo/common/URL.java
index 2e1eb3a17a..29d9df99a8 100644
--- a/dubbo-common/src/main/java/org/apache/dubbo/common/URL.java
+++ b/dubbo-common/src/main/java/org/apache/dubbo/common/URL.java
@@ -138 +138 @@ class URL implements Serializable {
-        this.urlParam = null;
+        this.urlParam = URLParam.parse(new HashMap<>());
@@ -148 +148 @@ class URL implements Serializable {
-        this.urlParam = urlParam;
+        this.urlParam = null == urlParam ? URLParam.parse(new HashMap<>()) : urlParam;
diff --git a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java b/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java
index 757f33345e..8e8b2f9424 100644
--- a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java
+++ b/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java
@@ -58 +58 @@ public class InstanceAddressURL extends URL {
-    
+
@@ -244,0 +245,4 @@ public class InstanceAddressURL extends URL {
+        if (null == serviceInfo) {
+            return getParameter(key);
+        }
+
@@ -309,0 +314,4 @@ public class InstanceAddressURL extends URL {
+        if (null == serviceInfo) {
+            return false;
+        }
+
@@ -346,0 +355,4 @@ public class InstanceAddressURL extends URL {
+        if (null == serviceInfo) {
+            return false;
+        }
+
@@ -430 +442,5 @@ public class InstanceAddressURL extends URL {
-       getServiceInfo(protocolServiceKey).addParameter(key, value);
+        MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+        if (null != serviceInfo) {
+            serviceInfo.addParameter(key, value);
+        }
+
@@ -439 +455,5 @@ public class InstanceAddressURL extends URL {
-        getServiceInfo(protocolServiceKey).addParameterIfAbsent(key, value);
+        MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+        if (null != serviceInfo) {
+            serviceInfo.addParameterIfAbsent(key, value);
+        }
+
@@ -444 +464,5 @@ public class InstanceAddressURL extends URL {
-        getServiceInfo(protocolServiceKey).addConsumerParams(params);
+        MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+        if (null != serviceInfo) {
+            serviceInfo.addConsumerParams(params);
+        }
+
@@ -459 +483,6 @@ public class InstanceAddressURL extends URL {
-            for (String fullKey : getServiceInfo(protocolServiceKey).getAllParams().keySet()) {
+            MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+            if (null == serviceInfo) {
+                return null;
+            }
+
+            for (String fullKey : serviceInfo.getAllParams().keySet()) {
@@ -480 +509,3 @@ public class InstanceAddressURL extends URL {
-        return getServiceInfo(protocolServiceKey).getNumbers();
+        MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+
+        return null == serviceInfo ? new ConcurrentHashMap<>() : serviceInfo.getNumbers();
@@ -497 +528,2 @@ public class InstanceAddressURL extends URL {
-        return getServiceInfo(protocolServiceKey).getMethodNumbers();
+        MetadataInfo.ServiceInfo serviceInfo = getServiceInfo(protocolServiceKey);
+        return null == serviceInfo ? new ConcurrentHashMap<>() : serviceInfo.getMethodNumbers();
diff --git a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java b/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java
index 35995f3c45..bfe1b8b383 100644
--- a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java
+++ b/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java
@@ -26,0 +27 @@ import org.apache.dubbo.common.utils.NetUtils;
+import org.apache.dubbo.metadata.MetadataInfo;
@@ -327 +328 @@ public class ServiceDiscoveryRegistryDirectory<T> extends DynamicDirectory<T> {
-            if(!(oldURL instanceof OverrideInstanceAddressURL && newURL instanceof OverrideInstanceAddressURL)) {
+            if (!(oldURL instanceof OverrideInstanceAddressURL && newURL instanceof OverrideInstanceAddressURL)) {
@@ -337,2 +338,6 @@ public class ServiceDiscoveryRegistryDirectory<T> extends DynamicDirectory<T> {
-        return !oldURL.getMetadataInfo().getValidServiceInfo(getConsumerUrl().getProtocolServiceKey())
-            .equals(newURL.getMetadataInfo().getValidServiceInfo(getConsumerUrl().getProtocolServiceKey()));
+        MetadataInfo.ServiceInfo oldServiceInfo = oldURL.getMetadataInfo().getValidServiceInfo(getConsumerUrl().getProtocolServiceKey());
+        if (null == oldServiceInfo) {
+            return false;
+        }
+
+        return !oldServiceInfo.equals(newURL.getMetadataInfo().getValidServiceInfo(getConsumerUrl().getProtocolServiceKey()));
